cpp/ql/src/semmle/code/cpp/controlflow/BackEdges.qll

```diff
@@ -0,0 +1,138 @@
+/**
+ * Provides a `backEdge` predicate, which can be useful when looking for
+ * cycles in the control flow graph.
+ *
+ * The implementation works by traversing the abstract syntax tree and
+ * assigning "entry" and "exit" indices to every element (corresponding
+ * roughly to pre- and post-order numberings).  We then choose an index for
+ * every element: usually the exit index, but for some nodes types, such as
+ * `ReturnStmt`, the results are better if we use the entry index instead.
+ * We then classify an edge as a back-edge if the index of the destination
+ * is less than or equal to the index of the source.
+ */
+
+import cpp
+
+/**
+ * Holds if there is a parent-child relationship between `parent` and
+ * `child`. For some nodes, we tinker with the order of the children to get
+ * indices that correspond better to the normal control flow order. For
+ * example, we want the condition of a DoStmt to get a higher index the
+ * loop body.
+ */
+private predicate parents(Element parent, int i, Element child) {
+  if parent instanceof ForStmt then (
+    exists (ForStmt loop
+    | loop = parent
+    | (child = loop.getInitialization() and i = 0) or
+      (child = loop.getCondition() and i = 1) or
+      (child = loop.getStmt() and i = 2) or
+      (child = loop.getUpdate() and i = 3))
+  ) else if parent instanceof DoStmt then (
+    exists (DoStmt loop
+    | loop = parent
+    | (child = loop.getStmt() and i = 0) or
+      (child = loop.getCondition() and i = 1))
+  ) else if parent instanceof AssignExpr then (
+    exists (Assignment assign
+    | assign = parent
+    | (child = assign.getRValue() and i = 0) or
+      (child = assign.getLValue() and i = 1))
+  ) else if parent instanceof FunctionCall then (
+    exists (FunctionCall call
+    | call = parent
+    | child = call.getArgument(i) or
+      (child = call.getQualifier() and i = call.getNumChild() - 1))
+  ) else if parent instanceof ExprCall then (
+    exists (ExprCall call
+    | call = parent
+    | child = call.getArgument(i) or
+      (child = call.getExpr() and i = call.getNumChild() - 1))
+  ) else (
+    exprparents(child,i,parent) or
+    stmtparents(child,i,parent) or
+    stmt_decl_bind(parent,i,child) or
+    (initialisers(child,parent,_,_) and i = 0)
```

Comment:
I really like the idea of building this on top of your QL CFG work. The purpose of this predicate is really to define the order of evaluation within an expression tree. It makes sense that you have written almost the same logic for the CFG work. So it would be much better to reuse your logic here.